### PR TITLE
Feat: desktopApi handshake

### LINKED ISSUE #4809: suite-desktop: create desktopApi.init message, pass all electron variables to renderer and store them in redux
Instead of calling separately `desktopApi.getHttpReceiverAddress`, `desktopApi.getUserDataInfo`, `desktopApi.getResourcesPath` (not exists yet but required by [trezor-connect-ipc](https://github.com/trezor/trezor-suite/pull/4704/) uploadFirmware method) we can create `desktopApi.init` message and send all paths/addresses once at the application startup.

### LINKED ISSUE #5627: create desktopApi.handshake flow
Extends https://github.com/trezor/trezor-suite/issues/4809

Desired app workflow:
1. create BrowserWindow (src-electron/app.ts init function) and start **only** hang-detect module, nothing else
2. renderer is loaded, call desktopApi.handshake [1]* and wait for the response (suite-desktop/src/index window.onload) consider rendering only lightweight preloader component without whole redux stack while waiting
3. handle handshake invocation (src-electron/app.ts)
   - clear timeout in hang-detect module
   - get system data (move this part from src-electron/app.ts init function)
   - loadModules (move this part from src-electron/app.ts init function) [2]*
   - respond with success or error
4. handle response in renderer, call store.dispatch action [3]* with data from handshake, handle action in reducer(s) **or** display error
5. render the whole thing (Main component)

---

[1] in the future we can preload indexedDB before the handshake and use stored data as a parameter of handshake method (like pass stored tor or autoupdate settings) #5534 this way we can get rid of "parallel storage", source of truth will be only in indexedDB (tor url, port, enabled etc)
[2] - it may be even faster/better if we split handshake into two steps: handshake1 - get only system data (to get "theme" as fast as we can to prevent "blink"), handshake2 - call `loadModules` (start tor, bridge etc)
[3] - in the future instead of dispatching action we can extend preloaded data from indexedDB and use them as initialState in `store` creation #5534

---

Problems/steps:
- [ ] we want to know some information about the system paths, theme etc before first render. related to https://github.com/trezor/trezor-suite/issues/4809 and  https://github.com/trezor/trezor-suite/pull/5534
- [ ] electron main process is starting up own subprocesses (tor, bridge etc) before it even knows that the app (renderer) is loaded properly. may be related to https://github.com/trezor/trezor-suite/issues/5612
- [ ] use data from the handshake (resourcePath) to get rid of [this workaround](https://github.com/trezor/trezor-suite/blob/develop/packages/suite-desktop/src-electron/modules/trezor-connect-ipc.ts#L72-L76=), set `basePath` in `TrezorConnect.firmwareUpdate` call (firmwareActions) instead
- [ ] get rid of `desktopApi.getUserDataInfo` function and module  (data should be present in handshake response > reducer, same case as firmwareUpdate)
- [ ] get rid of `desktopApi.clientReady` function (functionality replaced by handshake)

---
Probably follow up:
- [ ] finish state preloading #5534
- [ ] refactor auto-updater module and related events (info about the update may be sent in handshake response)
- [ ] find and refactor all others `desktopApi` calls from `useEffect` across suite components (tor/status, autoupdate, etc)

---
### QA ###
Please read [What to test](https://github.com/trezor/trezor-suite/pull/5632#user-content-qa) section in PR :heart:

### PR DESCRIPTION
**Pull request by szymonlesisz**
close https://github.com/trezor/trezor-suite/issues/5627, close https://github.com/trezor/trezor-suite/issues/4809
partial https://github.com/trezor/trezor-suite/issues/5118, related to https://github.com/trezor/trezor-suite/pull/5534

Read more [in the issue](https://github.com/trezor/trezor-suite/issues/5627)

### Prerequisites
- unifying entrypoint structure (`index.ts` and `Main.tsx`) of `suite-desktop` and `suite-web` (https://github.com/trezor/trezor-suite/pull/5632/commits/4f9f4cd5c72e38fcd4d516967a6557e6ec9a0903), should be pure refactoring
- wrapping `addToast` and `saveCoinProtocol` actions into a single `handleProtocolRequest` action (https://github.com/trezor/trezor-suite/pull/5632/commits/e84b9964123c6e7a5f22ad9eddf270b2c3c515e4) for later usage, should be pure refactoring

### Architectural changes
Electron _modules_ (modular code inside `suite-desktop/src-electron/modules`) are now 2-phased (https://github.com/trezor/trezor-suite/pull/5632/commits/9172fb2b066319e9e92e52908eb3afa3d311354c):
1. strictly synchronous `ModuleInit` functions, which may (or may not) return
1. asynchronous `ModuleLoad` functions, returning any result necessary.

The reason is that the `ModuleInit` stages are called before the renderer is prepared to display anything, therefore waiting would be very user-unfriendly. Then the renderer is started (by the modified `hangDetect` function) and after it responds with new `handshake/client` invocation, `ModuleLoad` stages are awaited (possibly with `HandshakeClient` payload from renderer, currently unused) and their outputs (`HandshakeElectron`) are sent back to the renderer as a `handshake/client` response. After that, renderer has everything it needs to start rendering the application.

While loading modules, `handshake/event` events are sent to the renderer which could e.g. show the progress (https://github.com/trezor/trezor-suite/pull/5632/commits/cc24f2c5447cbe8654d30398169b1e4c9f5fd78c, **DRAFT**).

Moreover, electron entrypoint `app.ts` was spaghettized for better readability (https://github.com/trezor/trezor-suite/pull/5632/commits/5d3380fac83a1321051ef61b41e7d6c2599001ae).

### Potential controversies
- ~as `desktopApi.getHttpReceiverAddress` was removed (replaced with `httpReceiver` url in handshake), `window.getHttpReceiverAddress` was created instead (https://github.com/trezor/trezor-suite/pull/5632/commits/411b76937fd4f4d37485d232a960d5a581f898d3). Its only purpose is not having to propagate the url from Redux to every utility function without access to Redux. Could probably be resolved in a more elegant way.~ `desktopApi.getHttpReceiverAddress` was preserved and should be refactored in the future.

### <a name="qa"></a>What to test
- only briefly that web Suite works as before, but most of the changes should affect only the desktop Suite
- that there is no new issue during application start
- opening `bitcoin:` URIs with application both opened/closed (on Win/Mac/Linux if possible)
- updating application (update notification, installation, etc. etc.)
- wiping user data
- Dropbox/Google (everything oauth related), because of http server
- everything Coinmarket related, because of http server
- firmware updating
- starting with Tor

### Followups
- discuss how to resolve `tor` and `bridge` modules, which are still not awaited because it would slow the application initalization too much.
- merge `desktopUpdateReducer` with the newly created `desktopReducer` which should process all the desktop-related stuff and remain `null` in web Suite.
- discuss creating `desktopMiddleware` which would automatically process desktop-emitted events and dispatch corresponding actions (without needing React stuff like `useTor` or `DesktopUpdater`)
- continue with #5118

### CODE REVIEW

File: packages/suite-desktop/src-electron/modules/http-receiver.ts [hunk 1]
```diff
@@ -49,6 +44,9 @@ const init: Module = ({ mainWindow }) => {
         logger.info('http-receiver', 'Stopping server (app quit)');
         httpReceiver.stop();
     });
+
+    // TODO check whether it's necessary to await httpReceiver.start()
+    return () => httpReceiver.getInfo();
```
Comment (mroz22) [inline packages/suite-desktop/src-electron/modules/http-receiver.ts]: I'd say it should be awaited, yes. It seems like a possble source of raceconditions.

Comment (szymonlesisz) [reply packages/suite-desktop/src-electron/modules/http-receiver.ts]: true, i did catch it here :)
turns out `libs/http-receiver` class require a little bit more love :) 
currently it was fired up immediately from the constructor which was declared on top level of `http-receiver` module, now whole initialization it is moved into "load" function

**Review by szymonlesisz — COMMENTED** [review]
looks good and i would approve it but i cant since i created this PR :)

didnt test everything. think this require a closer look from @trezor/qa regarding:
- custom protocols
- autoupdate
- theme
- stability of loading/reloading desktop app

**Review by mroz22 — APPROVED** [review]

**Comment by bosomt** [comment]
**QA OK**

so far we have not found any kind of problem in Suite app /bitcoin uri,update modal, wiping data, Spend, Tor, Start with tor, FW update, app restart, app reload, 3x labeling, 2fa/ we will continue with overall app testing after todays freeze

i only noticed one warning i havent seen before, not sure if its related to this change
`2022-08-02T07:08:47.932Z - WARN(http-receiver): Route not found for /favicon.ico`

**Info:**
 - Suite version: desktop 22.8.0 (d4346b1a7747383dc7e08fc479c146a9ca7e5fa9)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/22.8.0 Chrome/100.0.4896.160 Electron/18.3.5 Safari/537.36
 - OS: MacIntel
 - Screen: 1680x1050
 - Device: model T 2.5.2 regular (revision 426eae4dfcb853041419030bbb104cc2bbf6629f)
 - Transport: bridge 2.0.31

**Pull request merged**